lib/opencensus/tags/tag_map.rb

```diff
@@ -1,71 +1,116 @@
+    # Collection of tags that can be used to label anything that is
+    # associated with a stats or trace operations.
+    #
     # @example
     #
-    #  tag_map = OpenCensus::Tags::OpenCensus.new
+    #  tag_map = OpenCensus::Tags::TagMap.new
+    #
+    #  # Add tag
+    #  tag_map << OpenCensus::Tags::Tag.new "key1", "val1"
     #
-    #  # Add or update
-    #  tag_map["key1"] = "value1"
-    #  tag_map["key2"] = "value2"
+    #  # Add tag with TagTTL
+    #  tag_map << OpenCensus::Tags::Tag.new "key2", "val2", ttl: 0
     #
-    #  # Get value
-    #  tag_map["key1"] # value1
+    #  # Update tag
+    #  tag_map << OpenCensus::Tags::Tag.new "key3", "val3"
+    #  tag_map << OpenCensus::Tags::Tag.new "key3", "updatedval3"
+    #
+    #  # Get tag by key
+    #  p tag_map["key1"]
+    #  # <OpenCensus::Tags::Tag:0x007ffc138 @key="key1", @value="val1", @ttl=-1>
     #
     #  # Delete
     #  tag_map.delete "key1"
     #
     #  # Iterate
-    #  tag_map.each do |key, value|
-    #    p key
-    #    p value
+    #  tag_map.each do |tag|
+    #    p tag.key
+    #    p tag.value
     #  end
     #
     #  # Length
     #  tag_map.length # 1
     #
-    # @example Create tag map from hash
+    # @example Create tag map with list of tags.
     #
-    #   tag_map = OpenCensus::Tags::OpenCensus.new({ "key1" => "value1"})
+    #   tag_map = OpenCensus::Tags::OpenCensus.new([
+    #     OpenCensus::Tags::Tag.new "key1", "val1",
+    #     OpenCensus::Tags::Tag.new "key2", "val2"
+    #   ])
     #
     class TagMap
       extend Forwardable
 
-      # The maximum length for a tag key and tag value
-      MAX_LENGTH = 255
-
-      # Create a tag map. It is a map of tags from key to value.
-      # @param [Hash{String=>String}] tags Tags hash with string key and value.
+      # Create a tag map.
       #
-      def initialize tags = {}
-        @tags = {}
+      # @param [Array<Tags::Tag>] tags Tags list with string key and value and
+      #   metadata.
+      #
+      def initialize tags = []
+        @tags = tags.each_with_object({}) { |tag, r| r[tag.key] = tag }
+      end
 
-        tags.each do |key, value|
-          self[key] = value
-        end
+      # Insert tag.
+      #
+      # @param [Tag] tag
+      def << tag
+        @tags[tag.key] = tag
       end
 
-      # Set tag key value
+      # Get all tags
       #
-      # @param [String] key Tag key
-      # @param [String] value Tag value
-      # @raise [InvalidTagError] If invalid tag key or value.
+      # @returns [Array<Tag>]
+      def tags
+        @tags.values
+      end
+
+      # Get tag by key
       #
-      def []= key, value
-        validate_key! key
-        validate_value! value
+      # @return [Tag, nil]
+      def [] key
+        @tags[key]
+      end
 
-        @tags[key] = value
+      # Delete tag by key
+      # @param [String] key Tag key
+      def delete key
+        @tags.delete key
       end
 
+      # @!method each
+      #   @see Array#each
```

Comment:
Probably should be `Map#each` since `@tags` is a map (and its `#each` behavior will yield both the key and value)

Reply:
Iterating over `@tags` values which are `Tag` object is better because key is already in Tag instance with other data.

Structure of TagMap
```
{ String(tag key) => Tag instance(key, value, metadata)
```